### ISSUE TITLE
Getting where clauses?

### ISSUE THREAD

**Issue by mdiep** [jpsim/SourceKitten#554]
Given this file:

```swift
func f<C: Collection>(_: C) where C.Element == Int {}
```

Is there any way to get the `C.Element == Int` information?

When I run `sourcekitten structure`, nothing like that is included:

```swift
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.length" : 54,
  "key.offset" : 0,
  "key.substructure" : [
    {
      "key.accessibility" : "source.lang.swift.accessibility.internal",
      "key.bodylength" : 0,
      "key.bodyoffset" : 52,
      "key.kind" : "source.lang.swift.decl.function.free",
      "key.length" : 53,
      "key.name" : "f(_:)",
      "key.namelength" : 22,
      "key.nameoffset" : 5,
      "key.offset" : 0,
      "key.substructure" : [
        {
          "key.elements" : [
            {
              "key.kind" : "source.lang.swift.structure.elem.typeref",
              "key.length" : 10,
              "key.offset" : 10
            }
          ],
          "key.inheritedtypes" : [
            {
              "key.name" : "Collection"
            }
          ],
          "key.kind" : "source.lang.swift.decl.generic_type_param",
          "key.length" : 13,
          "key.name" : "C",
          "key.namelength" : 1,
          "key.nameoffset" : 7,
          "key.offset" : 7
        },
        {
          "key.kind" : "source.lang.swift.decl.var.parameter",
          "key.length" : 4,
          "key.namelength" : 0,
          "key.nameoffset" : 0,
          "key.offset" : 22,
          "key.typename" : "C"
        }
      ]
    }
  ]
}
```

I'm just getting started with SourceKit/SourceKitten and would appreciate any direction you could give or resources I should be looking at. 😅

**Comment by norio-nomura** [comment]
There is no way to directly extract `where…`, so we need to combine some informations and retrieve it.
```json
      "key.bodylength" : 0,
      "key.bodyoffset" : 52,
      "key.kind" : "source.lang.swift.decl.function.free",
      "key.length" : 53,
      "key.name" : "f(_:)",
      "key.namelength" : 22,
      "key.nameoffset" : 5,
      "key.offset" : 0,
```
Between `offset` and `bodyoffset`, finding offset of `where` by using `syntax` output.
`syntaxtype` is `source.lang.swift.syntaxtype.keyword` and content is `where`.
```json
  {
    "length" : 5,
    "offset" : 28,
    "type" : "source.lang.swift.syntaxtype.keyword"
  },
```
By using above, it might be assumed there is `where…` at 28..<52

**Comment by mdiep** [comment]
Is this a limitation of SourceKit, then? I imagine SourceKitten would pass through all information it receives.

**Comment by jpsim** [comment]
Correct, SourceKitten doesn't filter _out_ anything from SourceKit's responses.

However, there might be other SourceKit requests that can get you closer to what you need:

```json
{
  "key.accessibility" : "source.lang.swift.accessibility.internal",
  "key.annotated_decl" : "<Declaration>func f&lt;C&gt;(_: <Type usr=\"s:10YourModule1fyxs10CollectionRzSi7ElementRtzlF1CL_xmfp\">C<\/Type>) where C : <Type usr=\"s:s10CollectionP\">Collection<\/Type>, C.Element == <Type usr=\"s:Si\">Int<\/Type><\/Declaration>",
  "key.bodylength" : 0,
  "key.bodyoffset" : 52,
  "key.filepath" : "Sources\/YourModule\/Func.swift",
  "key.fully_annotated_decl" : "<decl.function.free><syntaxtype.keyword>func<\/syntaxtype.keyword> <decl.name>f<\/decl.name>&lt;<decl.generic_type_param usr=\"s:10YourModule1fyxs10CollectionRzSi7ElementRtzlF1CL_xmfp\"><decl.generic_type_param.name>C<\/decl.generic_type_param.name><\/decl.generic_type_param>&gt;(<decl.var.parameter><decl.var.parameter.argument_label>_<\/decl.var.parameter.argument_label>: <decl.var.parameter.type><ref.generic_type_param usr=\"s:10YourModule1fyxs10CollectionRzSi7ElementRtzlF1CL_xmfp\">C<\/ref.generic_type_param><\/decl.var.parameter.type><\/decl.var.parameter>) <syntaxtype.keyword>where<\/syntaxtype.keyword> <decl.generic_type_requirement>C : <ref.protocol usr=\"s:s10CollectionP\">Collection<\/ref.protocol><\/decl.generic_type_requirement>, <decl.generic_type_requirement>C.Element == <ref.struct usr=\"s:Si\">Int<\/ref.struct><\/decl.generic_type_requirement><\/decl.function.free>",
  "key.kind" : "source.lang.swift.decl.function.free",
  "key.length" : 53,
  "key.name" : "f(_:)",
  "key.namelength" : 22,
  "key.nameoffset" : 5,
  "key.offset" : 0,
  "key.parsed_declaration" : "func f<C: Collection>(_: C) where C.Element == Int",
  "key.parsed_scope.end" : 1,
  "key.parsed_scope.start" : 1,
  "key.substructure" : [

  ],
  "key.typename" : "<C where C : Collection, C.Element == Int> (C) -> ()",
  "key.typeusr" : "_T0yxcs10CollectionRzSi7ElementRtzluD",
  "key.usr" : "s:10YourModule1fyxs10CollectionRzSi7ElementRtzlF"
}
```

Specifically the `"key.fully_annotated_decl"` and `"key.fully_annotated_decl"` values might have useful information.

The main tradeoff here is that `structure` doesn't require the compiler arguments needed to compile the file(s) you're processing, but the command that generated the output above does.

If the output above has the information you need, you can use `sourcekitten doc` to process your files. See https://github.com/jpsim/SourceKitten#doc for sample uses.

**Comment by mdiep** [comment]
Why does `key.fully_annotated_decl` come across as an XML structure and not a JSON substructure? (Sorry for all the questions. I've been finding it hard to find explanations for all these things.)

**Comment by jpsim** [comment]
Because that's how SourceKit sends it back in the response. I've also found the rationales for how SourceKit does most things a bit hard to infer.

Also a word of caution, SourceKit doesn't seem to use standard JSON or XML encoding or decoding libraries, so you'll occasionally encounter malformed serializations of those encodings. SourceKitten has some sanitization for a few cases built in IIRC, and we've tried to file Swift JIRAs whenever possible to report issues.
